### PR TITLE
Tekton prow scripts

### PR DESCRIPTION
**Pull request by gabrielwen**
Implements Tekton runner to run Tekton pipelines.  Also provides examples of Task in templates and how to link from Prow to the tests in prow_config.yaml and PipelineRun configs.

Related to #425 

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/testing/622)

### CODE REVIEW

**Comment by gabrielwen** [comment]
progress as of today: steps in a task will stop running if any of previous steps has failed.  this is a WIP feature in Tekton to solve.  So we still need to use PVC to pass artifacts to another task for copy to jenkins GCS.  hopefully this will cause less problems as the files are small.  still, within a DAG if previous task fails, the following task doesn't seem to be running as well.  need to find a way to enforce it.

**Comment by gabrielwen** [comment]
@jlewi this is ready.

**Comment by gabrielwen** [comment]
@jlewi not yet - failed test junit xml is not able to upload to GCS, looking into it.

**Comment by gabrielwen** [comment]
@jlewi ok it's ready for review.

**Comment by gabrielwen** [comment]
hold on, I just found out I need to update the workflow to work with post-submits and periodic tests.

**Comment by jlewi** [comment]
Really great to see this coming along.

Could you call out in the PR description some of the limitations; e.g.
i) the need for PVC to work around that a failed step will block subsequent steps and tasks
   * Can we eliminate that by using Tekton's GCS resource to upload the artifacts from /workspace?

ii) The work around for teardown steps

File: py/kubeflow/testing/run_e2e_workflow.py [hunk 1]
```diff
@@ -343,6 +350,86 @@ def run(args, file_handler): # pylint: disable=too-many-statements,too-many-bran
...
+        test_target_name = workflow_name
+      logging.info("Reading Tekton PipelineRun config: %s", test_target_name)
+      tekton_run["metadata"]["name"] = workflow_name
+
+      for t in tekton_run.get("spec", {}).get("pipelineSpec", {}).get("tasks", []):
+        if not "params" in t:
+          t["params"] = []
+
+        # Update Junit artifacts XML.
+        for i in range(len(t["params"])):
```
Comment (jlewi) [inline py/kubeflow/testing/run_e2e_workflow.py]: Could you explain why run_e2e_workflow.py has to adjust the path of the junit files?

File: tekton/configs/junit-pvc.yaml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+apiVersion: v1
```
Comment (jlewi) [inline tekton/configs/junit-pvc.yaml]: Is it 1 PVC per task? per pipeline run?

File: tekton/templates/debug-task.yaml [hunk 3]
```diff
@@ -0,0 +1,87 @@
+apiVersion: tekton.dev/v1alpha1
```
Comment (jlewi) [inline tekton/templates/debug-task.yaml]: Could you add a comment explaining what this task is for?

File: tekton/templates/notebook-test-task.yaml [hunk 4]
```diff
@@ -8,6 +6,8 @@ metadata:
   annotations:
     sidecar.istio.io/inject: "false"
 spec:
+  workspaces:
+  - name: junit-artifacts
```
Comment (jlewi) [inline tekton/templates/notebook-test-task.yaml]: What are Tekton workspaces and how are they being used?

File: tekton/templates/test-teardown-task.yaml [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    - name: build-id
+      type: string
+      description:
+        BUILD_ID from Prow ENV.
+    resources:
+    - name: kf-testing-repo
+      type: git
+      targetPath: src/kubeflow/testing
+  steps:
+  - name: copy-artifacts
```
Comment (jlewi) [inline tekton/templates/test-teardown-task.yaml]: Can we use Tekton resources to upload files to GCS as opposed to having a separate task for that?

**Comment by jlewi** [comment]
@gabrielwen let me know when this is ready for another look.

File: tekton/templates/test-teardown-task.yaml (hunk 5 above)
Comment (gabrielwen) [reply tekton/templates/test-teardown-task.yaml]: just confirmed we could.

**Comment by gabrielwen** [comment]
@jlewi I think I got what we need for the design.  let's discuss in the doc and update here later on when we have a concrete version.  will share the doc in 1 - 2 days.

**Comment by jlewi** [comment]
@gabrielwen What is the status of this? When do you think it will be ready for review?

**Comment by gabrielwen** [comment]
@jlewi this is ready.  however, it seems all the tests are failing.  could you take a look and force merge it when it's ready?

**Comment by jlewi** [comment]
@gabrielwen Could you update the PR description to explain what this PR is doing? Use the PR description to help the reviewer figure out what they are looking at and what they should focus on.

**Comment by jlewi** [comment]
This is really great. Excited to get this in.

**Comment by jlewi** [comment]
@gabrielwen Any chance when another round will be ready for review?

**Comment by jlewi** [comment]
/lgtm
/approve
I think we can merge this as is and address the rest of the comments in follow on PRs. 

It looks like the Tekton test for mnist is failing. We probably shouldn't be running that test for kubeflow/testing anyway. A lint or unittest would probably make more sense. Maybe we can just remove those tests.

**Comment by jlewi** [comment]
/lgtm
/approve

Tests will probably fail because of #659 once that is fixed we should be able to merge this.

**Pull request merged**